test-framework/junit5/src/main/java/io/quarkus/test/junit/ClassloaderApplicationClassPredicate.java

```diff
@@ -0,0 +1,24 @@
+package io.quarkus.test.junit;
+
+import java.util.List;
+
+import javax.inject.Singleton;
+
+import io.quarkus.bootstrap.classloading.ClassPathElement;
+import io.quarkus.bootstrap.classloading.QuarkusClassLoader;
+import io.quarkus.runtime.test.TestApplicationClassPredicate;
+
+@Singleton
+class ClassloaderApplicationClassPredicate implements TestApplicationClassPredicate {
+
+    @Override
+    public boolean test(String name) {
+        QuarkusClassLoader cl = (QuarkusClassLoader) Thread.currentThread()
+                .getContextClassLoader();
+        //if the class file is present in this (and not the parent) CL then it is an application class
+        List<ClassPathElement> res = cl
+                .getElementsWithResource(name.replace(".", "/") + ".class", true);
+        return !res.isEmpty();
```

Comment:
Unfortunately, it seems that this approach does not work at all. I've copied this code from `IsolatedDevModeMain` but we probably need something different for tests. @stuartwdouglas Any idea?

Reply:
That won't work in tests, and TBH I don't know if the concept really makes sense. In dev mode 'application class' has a very specific meaning that is a class that can be hot reloadable. In tests I think this is a bit less clear. 

My original idea around this that I did not really articulate clearly was to basically just have a @DisableForTests annotation that you could apply to  the problematic observer method.

Reply:
> In tests I think this is a bit less clear.

My understanding of an _application class_ is that it's a class from the application root (`src/main/java`), or root of any other module in a multi-module project, or test dir (`src/test/java`), i.e. the classes that form the app except for classes coming from dependencies (quarkus extensions or libs).

Is the definition of _hot deployable resources_ different?

> My original idea around this that I did not really articulate clearly was to basically just have a @DisableForTests annotation that you could apply to the problematic observer method.

The main disadvantage of this approach is that the problematic observer would be always disabled (in all tests), whereas the approach in this PR is based on test profiles.

Reply:
Ok, so I've replaced the class loader approach with a bean "preloded" during STATIC_INIT. It works just fine, the question is whether we're OK with the concept in general...